texting_history/schema.py

```diff
@@ -0,0 +1,236 @@
+
+        if not is_request_user_verified(request):
+            raise GraphQLError("User must be verified via two-factor authentication!")
+
+        if not user.has_perm(VIEW_TEXT_MESSAGE_PERMISSION):
+            raise GraphQLError("User does not have permission to view text messages!")
+
+        return fn(parent, info, *args, **kwargs)
+
+    return wrapper
+
+
+@ensure_request_has_verified_user_with_permission
+def resolve_conversation(
+    parent,
+    info,
+    phone_number: str,
+    first: int,
+    after_or_at: float,
+) -> TextMessagesResult:
+    kwargs: Dict[str, Any] = {'user_phone_number': phone_number}
+    if after_or_at:
+        kwargs['ordering__lte'] = after_or_at
+    qs = Message.objects.filter(**kwargs).order_by('-ordering')
+    messages = list(qs[:first])
+    return TextMessagesResult(
+        messages=messages,
+        has_next_page=qs.count() > len(messages)
+    )
+
+
+@ensure_request_has_verified_user_with_permission
+def resolve_conversations(
+    parent,
+    info,
+    query: str,
+    first: int,
+    after_or_at: float,
+) -> LatestTextMessagesResult:
+    where_clauses: List[str] = []
+
+    with_clause = f"WITH latest_conversation_msg AS ({CONVERSATIONS_SQL_FILE.read_text()})"
+
+    select_with_user_info_statement = """
+    SELECT
+        msg.*,
+        usr.id as user_id,
+        usr.first_name || ' ' || usr.last_name as user_full_name
+    FROM
+        latest_conversation_msg AS msg
+    LEFT JOIN
+        users_justfixuser AS usr ON '+1' || usr.phone_number = msg.user_phone_number
+    """
+
+    if after_or_at:
+        where_clauses.append("(ordering <= %(after_or_at)s)")
+    if ALL_DIGITS_RE.fullmatch(query):
+        where_clauses.append("(user_phone_number LIKE '+1' || %(query)s || '%%')")
+    elif query:
+        where_clauses.append(
+            "((usr.first_name || ' ' || usr.last_name) ILIKE '%%' || %(query)s || '%%')"
+        )
+
+    where_clause = ('WHERE ' + ' AND '.join(where_clauses)) if where_clauses else ''
+    order_clause = "ORDER BY ordering DESC"
+    limit_clause = f"LIMIT %(first)s"
+    sql_args = {
+        'first': first,
+        'after_or_at': after_or_at,
+        'query': query,
+    }
+
+    with connection.cursor() as cursor:
+        base_select = '\n'.join([
+            select_with_user_info_statement,
+            where_clause,
+        ])
+        cursor.execute('\n'.join([
+            with_clause,
+            base_select,
+            order_clause,
+            limit_clause,
+        ]), sql_args)
+        messages = [LatestTextMessage(**row) for row in generate_json_rows(cursor)]
+        cursor.execute('\n'.join([
+            f"{with_clause}, filtered_messages AS ({base_select})",
+            'SELECT COUNT(*) FROM filtered_messages',
+        ]), sql_args)
+        count = cursor.fetchone()[0]
+        return LatestTextMessagesResult(
+            messages=messages,
+            has_next_page=count > len(messages)
+        )
```

Comment:
@sraby, I think these two functions illustrate the potential advantages of using Django's QuerySet API nicely.

In this case we want to be able to execute a query and get its rows, but we also want to be able to get the total count of rows, so that we can tell the client whether or not there's a next page of results they can fetch...  As far as I know, there's no way to request both of these things in a single SQL query.

The top function, `resolve_conversation()`, does this by using Django's QuerySet API to dynamically construct the query, and then is easily able to retrieve both a subset of the results of the query (via Python's slice notation, which is converted to a `LIMIT` clause in SQL) and a row count of the total results (via the QuerySet's `.count()` method, which takes the same query but wraps it in a SQL aggregation for us).

On the other hand, in the second function, `resolve_conversations()` (yeah, I'm not a fan of how similar the function names are), I had to manually cobble together a query, because the query I wanted to make wasn't possible (or at least wasn't easy) using Django's QuerySet API... you can see how relatively convoluted the code got as I had to set up two separate queries, one to retrieve the query rows and another to get the count.  I'm definitely not a QuerySet or SQL expert though, so there might have been a better way to do things that I'm not aware of.